# First Delivery
### Getting started

##### WAPXXX: 0 (no signal) 1 (has signal)
##### ID: Classroom identifyer

 - Are there enough samples for each ID class?
 - Do we have WAPXXX helping us to predict ID? Indicate 5 of them with more potential prediction and 5 of them with less.
 ---> Cleaned Dataset at: ../datasets/UJIIndoorLoc/UJIIndoorLoc_B0-ID-01-AfterDelivery1.csv

In [1]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv("../datasets/UJIIndoorLoc/UJIIndoorLoc_B0-ID-01.csv")

Checking the data characteristics

In [3]:
data.shape
data.describe()
data.info()
data.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5249 entries, 0 to 5248
Columns: 521 entries, WAP001 to ID
dtypes: int64(521)
memory usage: 20.9 MB


WAP001  WAP002  WAP003  WAP004  WAP005  WAP006  WAP007  WAP008  WAP009  \
0       0       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   

   WAP010  ...  WAP512  WAP513  WAP514  WAP515  WAP516  WAP517  WAP518  \
0       0  ...       0       0       0       0       0       0       0   
1       0  ...       0       0       0       0       0       0       0   

   WAP519  WAP520   ID  
0       0       0  122  
1       0       0  102  

[2 rows x 521 columns]

In [4]:
data.dtypes

WAP001    int64
WAP002    int64
WAP003    int64
WAP004    int64
WAP005    int64
          ...  
WAP517    int64
WAP518    int64
WAP519    int64
WAP520    int64
ID        int64
Length: 521, dtype: object

Droping columns with all 0

In [5]:
data = data.loc[:, (data!=0).any('index')]
data.shape

(5249, 201)

In [6]:
len(data['ID'].unique())

256

###### So we for this 5249 entries we have 256 unique ID's and 201 relevant columns after droping the zero ones thus they does not give us information.
###### Therefore there we can answer the first question clarifying that would be necessary to have more entries.

In [7]:
duplicated_data = data[data.duplicated(keep=False)]
duplicated_data.head(5)

WAP001  WAP002  WAP007  WAP008  WAP009  WAP013  WAP014  WAP017  WAP018  \
8         0       0       0       0       0       0       0       0       0   
35        0       0       1       0       0       1       1       0       0   
47        0       0       1       0       0       1       1       0       0   
196       0       0       0       0       0       0       0       0       0   
197       0       0       0       0       0       0       0       0       0   

     WAP019  ...  WAP447  WAP452  WAP459  WAP475  WAP494  WAP500  WAP508  \
8         0  ...       0       1       0       0       0       0       0   
35        0  ...       0       0       0       0       0       0       0   
47        0  ...       0       0       0       0       0       0       0   
196       0  ...       0       0       0       0       0       0       0   
197       0  ...       0       0       0       0       0       0       0   

     WAP515  WAP519    ID  
8         0       0   113  
35        0       0   211  
47        0       0   232  
196       0       0  3104  
197       0       0  3107  

[5 rows x 201 columns]

Checking duplicated data of an especific ID

In [8]:
duplicated_data_211 = duplicated_data.loc[(duplicated_data['ID']) == 211]
duplicated_data_211

WAP001  WAP002  WAP007  WAP008  WAP009  WAP013  WAP014  WAP017  WAP018  \
35         0       0       1       0       0       1       1       0       0   
290        0       0       1       0       0       1       1       0       0   
1776       0       0       1       0       0       1       1       0       0   
2033       0       0       1       0       0       1       1       0       0   
3323       0       0       0       0       0       1       1       0       0   
3598       0       0       0       0       0       1       1       0       0   
3877       0       0       0       0       0       1       1       0       0   
4148       0       0       0       0       0       1       1       0       0   
4427       0       0       0       0       0       1       1       0       0   
4705       0       0       0       0       0       1       1       0       0   
4981       0       0       0       0       0       1       1       0       0   

      WAP019  ...  WAP447  WAP452  WAP459  WAP475  WAP494  WAP500  WAP508  \
35         0  ...       0       0       0       0       0       0       0   
290        0  ...       0       0       0       0       0       0       0   
1776       0  ...       0       0       0       0       0       0       0   
2033       0  ...       0       0       0       0       0       0       0   
3323       0  ...       0       1       0       0       0       0       0   
3598       0  ...       0       1       0       0       0       0       0   
3877       0  ...       0       1       0       0       0       0       0   
4148       0  ...       0       1       0       0       0       0       0   
4427       0  ...       0       1       0       0       0       0       0   
4705       0  ...       0       1       0       0       0       0       0   
4981       0  ...       0       1       0       0       0       0       0   

      WAP515  WAP519   ID  
35         0       0  211  
290        0       0  211  
1776       0       0  211  
2033       0       0  211  
3323       0       0  211  
3598       0       0  211  
3877       0       0  211  
4148       0       0  211  
4427       0       0  211  
4705       0       0  211  
4981       0       0  211  

[11 rows x 201 columns]

As we can see we can have more than one kind of duplicated data per ID

In [9]:
duplicated_data.loc[(duplicated_data['ID']) == 211].sum()

WAP001       0
WAP002       0
WAP007       4
WAP008       0
WAP009       0
          ... 
WAP500       0
WAP508       0
WAP515       0
WAP519       0
ID        2321
Length: 201, dtype: int64

Example of querying with iloc function 

In [10]:
duplicated_data_211.iloc[:, :].query('WAP007 == 1')

WAP001  WAP002  WAP007  WAP008  WAP009  WAP013  WAP014  WAP017  WAP018  \
35         0       0       1       0       0       1       1       0       0   
290        0       0       1       0       0       1       1       0       0   
1776       0       0       1       0       0       1       1       0       0   
2033       0       0       1       0       0       1       1       0       0   

      WAP019  ...  WAP447  WAP452  WAP459  WAP475  WAP494  WAP500  WAP508  \
35         0  ...       0       0       0       0       0       0       0   
290        0  ...       0       0       0       0       0       0       0   
1776       0  ...       0       0       0       0       0       0       0   
2033       0  ...       0       0       0       0       0       0       0   

      WAP515  WAP519   ID  
35         0       0  211  
290        0       0  211  
1776       0       0  211  
2033       0       0  211  

[4 rows x 201 columns]

##### Dropping duplicated data

In [11]:
data = data.drop_duplicates(keep=False)
data.shape

(3502, 201)

Finally we save the cleaned data set

In [12]:
data.to_csv('../datasets/UJIIndoorLoc/UJIIndoorLoc_B0-ID-01-AfterDelivery1.csv')

##### Correlating data to check which data has more and less potential
Absolute value and rounding to 6 decimals

In [13]:
corr = data.corr().abs().round(decimals=6)
corr

WAP001    WAP002    WAP007    WAP008    WAP009    WAP013    WAP014  \
WAP001  1.000000  0.003649  0.030185  0.018543  0.025519  0.053119  0.044313   
WAP002  0.003649  1.000000  0.021316  0.007196  0.018021  0.056887  0.069248   
WAP007  0.030185  0.021316  1.000000  0.059536  0.028500  0.253630  0.281000   
WAP008  0.018543  0.007196  0.059536  1.000000  0.050333  0.100472  0.095302   
WAP009  0.025519  0.018021  0.028500  0.050333  1.000000  0.323138  0.298193   
...          ...       ...       ...       ...       ...       ...       ...   
WAP500  0.011540  0.008150  0.067422  0.022761  0.057000  0.114831  0.117686   
WAP508  0.003439  0.002429  0.020094  0.006784  0.016988  0.052235  0.065278   
WAP515  0.024596  0.006171  0.051051  0.017235  0.043159  0.089839  0.089110   
WAP519  0.002105  0.001486  0.012296  0.136309  0.010396  0.021639  0.021464   
ID      0.067691  0.050019  0.305186  0.025543  0.134600  0.292123  0.286517   

          WAP017    WAP018    WAP019  ...    WAP447    WAP452    WAP459  \
WAP001  0.148544  0.023798  0.021557  ...  0.005971  0.039079  0.004035   
WAP002  0.006529  0.006292  0.015224  ...  0.004217  0.027598  0.002849   
WAP007  0.054017  0.052057  0.114318  ...  0.034884  0.211163  0.023573   
WAP008  0.095427  0.234868  0.042519  ...  0.011777  0.057495  0.007958   
WAP009  0.045667  0.044010  0.341161  ...  0.069310  0.154177  0.019929   
...          ...       ...       ...  ...       ...       ...       ...   
WAP500  0.020652  0.009985  0.041520  ...  0.013337  0.112699  0.023592   
WAP508  0.006155  0.005932  0.014351  ...  0.003975  0.026016  0.002686   
WAP515  0.015637  0.015070  0.036459  ...  0.010098  0.053190  0.035774   
WAP519  0.073369  0.003630  0.008782  ...  0.002432  0.015920  0.001644   
ID      0.019001  0.021001  0.117661  ...  0.127400  0.219816  0.075544   

          WAP475    WAP494    WAP500    WAP508    WAP515    WAP519        ID  
WAP001  0.001215  0.042465  0.011540  0.003439  0.024596  0.002105  0.067691  
WAP002  0.000858  0.007761  0.008150  0.002429  0.006171  0.001486  0.050019  
WAP007  0.007097  0.064209  0.067422  0.020094  0.051051  0.012296  0.305186  
WAP008  0.002396  0.021677  0.022761  0.006784  0.017235  0.136309  0.025543  
WAP009  0.006000  0.054283  0.057000  0.016988  0.043159  0.010396  0.134600  
...          ...       ...       ...       ...       ...       ...       ...  
WAP500  0.002713  0.524956  1.000000  0.007682  0.498285  0.004701  0.079281  
WAP508  0.000809  0.007316  0.007682  1.000000  0.005817  0.001401  0.056033  
WAP515  0.002055  0.507899  0.498285  0.005817  1.000000  0.003560  0.057344  
WAP519  0.000495  0.004477  0.004701  0.001401  0.003560  1.000000  0.035066  
ID      0.004308  0.096876  0.079281  0.056033  0.057344  0.035066  1.000000  

[201 rows x 201 columns]

Stacking the data to manage it

In [14]:
stacked_corr = corr.stack()
stacked_corr

WAP001  WAP001    1.000000
        WAP002    0.003649
        WAP007    0.030185
        WAP008    0.018543
        WAP009    0.025519
                    ...   
ID      WAP500    0.079281
        WAP508    0.056033
        WAP515    0.057344
        WAP519    0.035066
        ID        1.000000
Length: 40401, dtype: float64

Removing the innecessary data of the correlation (ones)

In [15]:
stacked_corr = stacked_corr[stacked_corr.index.get_level_values(0) != stacked_corr.index.get_level_values(1)]
stacked_corr

WAP001  WAP002    0.003649
        WAP007    0.030185
        WAP008    0.018543
        WAP009    0.025519
        WAP013    0.053119
                    ...   
ID      WAP494    0.096876
        WAP500    0.079281
        WAP508    0.056033
        WAP515    0.057344
        WAP519    0.035066
Length: 40200, dtype: float64

#### 5 Most and less important representative APs

In [16]:
stacked_corr.sort_values(ascending=False).head(5)

WAP414  WAP363    1.000000
WAP363  WAP414    1.000000
WAP170  WAP171    1.000000
WAP171  WAP170    1.000000
WAP054  WAP053    0.892886
dtype: float64

In [17]:
stacked_corr.sort_values(ascending=True).head(5)

WAP415  WAP404    0.000000
WAP404  WAP415    0.000000
WAP447  WAP410    0.000010
WAP410  WAP447    0.000010
WAP289  WAP048    0.000054
dtype: float64